tests/ts/logger_helpers.ts

```diff
@@ -0,0 +1,511 @@
+ * Copyright © 2016-2019 Threema GmbH (https://threema.ch/).
+ *
+ * This file is part of Threema Web.
+ *
+ * Threema Web is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or (at
+ * your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with Threema Web. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+// tslint:disable:no-reference
+/// <reference path="../../src/threema.d.ts" />
+
+import {Logger} from 'ts-log';
+
+import {BaseConfidential} from '../../src/helpers/confidential';
+import {ConsoleLogger, LevelLogger, MemoryLogger, TagLogger, TeeLogger, UnveilLogger} from '../../src/helpers/logger';
+
+import LogLevel = threema.LogLevel;
+import LogType = threema.LogType;
+
+interface EnsureLoggedArgs {
+    level: LogLevel,
+    logger: TestLogger | TestLogger[],
+    tag?: string | string[],
+}
+
+type LogRecord = [LogType, any?, ...any[]];
+type LogFunction = (message?: any, ...args: any[]) => void;
+// Supported log (level) types
+const LOG_TYPES: LogType[] = ['debug', 'trace', 'info', 'warn', 'error'];
+
+
+/**
+ * Stores all log records in memory for evaluation.
+ */
+class TestLogger implements Logger {
+    public readonly records: LogRecord[] = [];
+    public readonly debug: LogFunction;
+    public readonly trace: LogFunction;
+    public readonly info: LogFunction;
+    public readonly warn: LogFunction;
+    public readonly error: LogFunction;
+
+    constructor() {
+        // Bind log level type methods
+        for (const type of LOG_TYPES) {
+            this[type] = this.append.bind(this, type);
+        }
+    }
+
+    private append(type: LogType, message?: any, ...args: any[]): void {
+        this.records.push([type, message, ...args])
+    }
+}
+
+/**
+ * A confidential log record for testing purposes.
+ */
+class TestConfidential extends BaseConfidential<string, string> {
+    public readonly uncensored: string = 'uncensored';
+
+    public censored(): string {
+        return 'censored';
+    }
+}
+
+/**
+ * Log a record of each type.
+ */
+function logEachType(...loggers: Logger[]): void {
+    for (const logger of loggers) {
+        for (const type of LOG_TYPES) {
+            logger[type](type);
+        }
+    }
+}
+
+/**
+ * Format a log record. This splices a tag into the expected place.
+ */
+function formatLogRecord(record: LogRecord, tag?: string[]): LogRecord {
+    if (tag === undefined) {
+        return record;
+    }
+
+    // Splice tag into message
+    const [type, ...args] = record;
+    return [type, ...tag.concat(args)];
```

Comment:
It took a few seconds for me to parse this... This is something where `record.tag = tag` would be *much* easier to read :slightly_smiling_face:

Reply:
This would require squeezing in-band meta-information through the interface defined by `ts-log` which I'm not willing to do.